lib/puma/cluster.rb

```diff
@@ -227,9 +227,11 @@ def check_workers(force=false)
       # during this loop by giving the kernel time to kill them.
       sleep 1 if any
 
-      @workers.reject! { |w| Process.waitpid(w.pid, Process::WNOHANG) }
-
-      @workers.reject!(&:dead?)
+      pids = []
+      while pid = Process.waitpid(-1, Process::WNOHANG) do
+        pids << pid
+      end
+      @workers.reject! { |w| w.dead? || pids.include?(w.pid) }
```

Comment:
Could move these lines into a `reap_workers` method to match `cull_workers` and `spawn_workers`.

Reply:
Good suggestion, I'll leave coding style up to otheers.  It's four line, probably only called by the method it's contained in, and any test would involve that method...